sklearn/metrics/_pairwise_distances_reduction.pyx

```diff
@@ -204,35 +227,38 @@ cdef class PairwiseDistancesReduction:
         self.X_n_chunks = X_n_full_chunks + (self.X_n_samples_remainder != 0)
         self.Y_n_chunks = Y_n_full_chunks + (self.Y_n_samples_remainder != 0)
 
-    def compute(
-        self,
-        str strategy=None,
-        bint return_distance=False,
-    ):
-        """Compute the pairwise distances and the reduction of vectors (rows) of X on Y.
+        if strategy is None:
+            strategy = get_config().get("pairwise_dist_parallel_strategy", 'auto')
 
-        Parameters
-        ----------
-        strategy : str, {'auto', 'parallel_on_X', 'parallel_on_Y'}, default=None
-            The chunking strategy defining which dataset parallelization are made on.
+        if strategy not in ('parallel_on_X', 'parallel_on_Y', 'auto'):
+            raise RuntimeError(f"strategy must be 'parallel_on_X, 'parallel_on_Y', "
+                               f"or 'auto', but currently strategy='{self.strategy}'.")
 
-            Strategies differs on the dispatching they use for chunks on threads:
+        if strategy == 'auto':
+            # This is a simple heuristic whose constant for the
+            # comparison has been chosen based on experiments.
+            if 4 * self.chunk_size * self.available_n_threads < self.n_samples_X:
+                strategy = 'parallel_on_X'
+            else:
+                strategy = 'parallel_on_Y'
 
-              - 'parallel_on__X' dispatches chunks of X uniformly on threads.
-              Each thread then iterates on all the chunks of Y. This strategy is
-              embarrassingly parallel and comes with no datastructures synchronisation.
+        self.execute_in_parallel_on_Y = strategy == "parallel_on_Y"
 
-              - 'parallel_on_Y' dispatches chunks of Y uniformly on threads.
-              Each thread then iterates on all the chunks of X. This strategy is
-              embarrassingly parallel but uses intermediate datastructures
-              synchronisation.
+        # Not using less, not using more.
+        self.effective_n_threads = min(
```

Comment:
For maintainability, I think we need a comment describing when to use `self.effective_n_threads` vs `self.available_n_threads`.

Reply:
9cda95f03e69ef9e04c2429ed233be11254e5e82 is an attempt addressing it.